### PR TITLE
Add EXIV2_ENABLE_FILESYSTEM_ACCESS option

### PR DESCRIPTION
**Pull request by enen92**
Hey,

I am trying to add exiv2 as a dependency for Kodi/XBMC as it seems the most complete GPL2 library for image metadata extraction (since it covers both EXIF and IPTC which we currently do with custom parsers - pretty much untouched since the original xbox days ~ 14 years ago). As you might easily guess, it's pretty much broken by now :)
 
Unfortunately, since the application is available on multiple platforms some of which are not supported by default by exiv2 (e.g. android, tvos, ios, xbox/uwp, etc) filesystem access is not something that can be taken for granted. The usage we plan to do of the library only requires that we are able to read the image (for tag extraction) from a memory buffer - hence we only really need the MemIO.

https://github.com/xbmc/xbmc/pull/24109/files#diff-40b93a8f1858865324006f152285b7b28de2a1b8e878cb9e04ac4f4f4d11dd0fR35-R36

PoC for inclusion is being done in https://github.com/xbmc/xbmc/pull/24109

This PR proposes an additional option `EXIV2_ENABLE_FILESYSTEM_ACCESS` to allow disable `FileIO` and generic filesystem access. This is only valid if unit tests, the exiv app, fuzzer, etc are not enable so I think it shouldn't affect the normal behaviour of the application.
In case this is accepted probably the CI should have a way to validate further builds.

### CODE REVIEW

File: src/makernote_int.cpp [hunk 1]
```diff
@@ -81,13 +84,15 @@ std::string getExiv2ConfigPath() {
...
+#endif
+  return "";
 }
 
 std::string readExiv2Config([[maybe_unused]] const std::string& section, [[maybe_unused]] const std::string& value,
                             const std::string& def) {
   std::string result = def;
 
-#ifdef EXV_ENABLE_INIH
+#if defined(EXV_ENABLE_INIH) and defined(EXV_ENABLE_FILESYSTEM)
```
Comment (kmilos) [inline src/makernote_int.cpp:95]:
```suggestion
#if defined(EXV_ENABLE_INIH) && defined(EXV_ENABLE_FILESYSTEM)
```

Comment (enen92) [reply src/makernote_int.cpp]: thanks, let's see if I finally get a 100% green this time :)

**Comment by enen92** [comment]
Added a new commit to add a new job to validate a build with filesystem disabled. This likely won't run as part of this PR but you can check the result here: https://github.com/enen92/exiv2/actions/runs/6937027229/job/18870288972?pr=2

**Comment by enen92** [comment]
rebased after https://github.com/Exiv2/exiv2/pull/2844 got merged

File: src/basicio.cpp [hunk 2]
```diff
@@ -47,12 +48,13 @@ namespace fs = std::filesystem;
...
 namespace fs = std::experimental::filesystem;
 #endif
+#endif
 
 // *****************************************************************************
 // class member definitions
 namespace {
 /// @brief replace each substring of the subject that matches the given search string with the given replacement.
-void ReplaceStringInPlace(std::string& subject, std::string_view search, std::string_view replace) {
+[[maybe_unused]] void ReplaceStringInPlace(std::string& subject, std::string_view search, std::string_view replace) {
```
Comment (neheb) [inline src/basicio.cpp:57]: probably makes sense to get rid of this as there's a single usage. Maybe convert to lambda.

File: src/makernote_int.cpp [hunk 3]
```diff
@@ -81,13 +84,15 @@ std::string getExiv2ConfigPath() {
   currentPath = std::string(pw ? pw->pw_dir : "");
 #endif
   return (currentPath / inifile).string();
+#endif
+  return "";
```
Comment (neheb) [inline src/makernote_int.cpp:88]: this will trigger a compiler warning. Should move it within an else block.

File: src/image.cpp [hunk 4]
```diff
@@ -817,12 +822,15 @@ BasicIo::UniquePtr ImageFactory::createIo(const std::string& path, [[maybe_unuse
...
 #endif
+#ifdef EXV_ENABLE_FILESYSTEM
   if (fProt == pFileUri)
     return std::make_unique<FileIo>(pathOfFileUrl(path));
   if (fProt == pStdin || fProt == pDataUri)
     return std::make_unique<XPathIo>(path);  // may throw
 
   return std::make_unique<FileIo>(path);
+#endif
+  return nullptr;
```
Comment (neheb) [inline src/image.cpp:833]: this will trigger a compiler warning. Should move it within an else block.

File: src/futils.cpp [hunk 5]
```diff
@@ -382,5 +388,7 @@ std::string getProcessPath() {
     return "unknown";
   }
 #endif
+#endif
+  return "unknown";
```
Comment (neheb) [inline src/futils.cpp:392]: this will trigger a compiler warning. Should move it within an else block.

File: src/futils.cpp [hunk 6]
```diff
@@ -231,7 +233,10 @@ bool fileExists(const std::string& path) {
   if (fileProtocol(path) != pFile) {
     return true;
   }
+#ifdef EXV_ENABLE_FILESYSTEM
   return fs::exists(path);
+#endif
+  return false;
```
Comment (neheb) [inline src/futils.cpp:239]: this will trigger a compiler warning. Should move it within an else block.

**Review by neheb — APPROVED** [review]

**Pull request merged**